### PR TITLE
Computer

### PR DESCRIPTION
**Pull request by Gabbendorf**

### CODE REVIEW

File: lib/computer.rb [hunk 1]
```diff
@@ -8,29 +8,33 @@ class Computer
...
-  def place_ship
-    while @ships_list.ships.size > 0
-      random_ship = select_random_ship
-      position = valid_position(select_random_position, random_ship.length)
-      random_ship.register_position(position[:x], position[:y], position[:orientation])
-      @grid.add_ship(random_ship)
-    end
+  def ship_placement
+    random_ship = generate_ship
+    position = generate_position(random_ship.length)
```
Comment (fabientownsend) [inline lib/computer.rb:15]: When I read _generate_ I expected a _computered_ value, but it seems that this value can be returned by a human as well. Maybe something call next_move, new_move would be more correct?

File: spec/game_spec.rb [hunk 2]
```diff
@@ -7,8 +7,7 @@
 require_relative '../lib/ships_list'
 require_relative '../lib/validated_ui'
 require_relative '../lib/computer'
-require_relative '../lib/player'
-
+require_relative '../lib/human_player'
```
Comment (fabientownsend) [inline spec/game_spec.rb:10]: There is a way to replace those `require_relative` with `require 'name_of_the_file'` if you are using RSpec

File: spec/computer_spec.rb [hunk 3]
```diff
@@ -13,24 +13,21 @@
...
-    ship3_name = grid.ships_placed[2].name
-    ship4_name = grid.ships_placed[3].name
-    ship5_name = grid.ships_placed[4].name
-    ship6_name = grid.ships_placed[5].name
-    all_ship_names = [ship1_name, ship2_name, ship3_name, ship4_name, ship5_name, ship6_name]
-    is_valid_name = all_ship_names.map {|name| possible_ships_name.include?(name)}
-
-    expect(is_valid_name).to eq([true] * 6)
-    expect(grid.ships_placed.size).to eq(6)
+    expect(possible_ships.keys.include?(ship.name)).to eq(true)
```
Comment (fabientownsend) [inline spec/computer_spec.rb]: expect do a comparison between two value, as a `if` here you can possibly write: `expect(possible_ships.keys.include?(ship.name))`

File: lib/validated_ui.rb [hunk 4]
```diff
@@ -4,18 +4,17 @@
...
 class ValidatedUi
 
-  def initialize(ui, ships_list, grid)
+  def initialize(ui, grid)
     @ui = ui
-    @ships_list = ships_list
     @grid = grid
   end
 
-  def selected_ship(player_name, list)
```
Comment (fabientownsend) [inline lib/validated_ui.rb:13]: What does list represent here? Couldn't be replaced with a more explicit name?

File: lib/computer.rb [hunk 5]
```diff
@@ -8,29 +8,33 @@ class Computer
...
 
-  def place_ship
-    while @ships_list.ships.size > 0
-      random_ship = select_random_ship
-      position = valid_position(select_random_position, random_ship.length)
-      random_ship.register_position(position[:x], position[:y], position[:orientation])
-      @grid.add_ship(random_ship)
-    end
+  def ship_placement
+    random_ship = generate_ship
```
Comment (fabientownsend) [inline lib/computer.rb:14]: Why no directly call `select_random_ship` here? I'm not sure that `generate_ship` is really needed

File: lib/computer.rb [hunk 6]
```diff
@@ -8,29 +8,33 @@ class Computer
...
-    validation_result = @grid.validate_placement(position, ship_length)
+  def generate_ship
+    select_random_ship
+  end
+
+  def generate_position(ship_length)
+    valid_position(ship_length)
+  end
+
+  def valid_position(ship_length)
```
Comment (fabientownsend) [inline lib/computer.rb:30]: I feel like this method has two responsibility, get a random move and validate this random move. Wouldn't it be easier to read if it was split into two different methods? Something like that:

``` ruby
def get_next_position(ship)
    position = random_position(ship)
    return position if available?(position)
    get_next_position(ship)
end
```

File: lib/game.rb [hunk 7]
```diff
@@ -15,31 +21,37 @@ def initialize(grid_display, ui, grid, ships_list, computer, validated_ui)
...
-      while @ships_list.ships.size > 0
-        ship = ships_owner.placement_move(ships_owner.name, @ships_list, @ui)
-        @grid.add_ship(ship)
-      end
+      ships_placer = HumanPlayer.new(@ui.ask_name_player1, @grid, @validated_ui, @ui, @ships_list)
+    end
+  end
+
+  def ships_placement(ships_placer)
+    while @ships_list.ships.size > 0
```
Comment (fabientownsend) [inline lib/game.rb]: What does `0` represent here?

File: lib/grid.rb [hunk 8]
```diff
@@ -15,6 +15,8 @@ def validate_placement(position, ship_length)
       :invalid_ship_position
     elsif ship_out_of_grid?( position[:x], position[:y], ship_length, position[:orientation]) == true
       :invalid_placement
+    elsif already_occupied?(position) == true
```
Comment (fabientownsend) [inline lib/grid.rb]: the `== true` statement isn't required here

**Review by fabientownsend — COMMENTED** [review]
👍

File: lib/validated_ui.rb (hunk 4 above)
Comment (Gabbendorf) [reply lib/validated_ui.rb:13]: You mean something like "ships_list" or like "ships_array"?

File: spec/game_spec.rb (hunk 2 above)
Comment (Gabbendorf) [reply spec/game_spec.rb:10]: You mean `require 'spec'` ?

**Pull request merged**